### PR TITLE
Button component

### PR DESCRIPTION
**Pull request by hakimelek**
Adds the Button component.

#19

### CODE REVIEW

File: src/BackgroundImage.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import styled from 'styled-components'
```
Comment (jxnblk) [inline src/BackgroundImage.js]: As we discussed in person, let's merge the BackgroundImage PR first and see if that clears up some of these unrelated changes

File: src/Button/Button.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import React from 'react'
```
Comment (jxnblk) [inline src/Button/Button.js]: Let's keep the folder hierarchy flat for now, which will help with imports and merging – I don't see much value in putting a module named `Button` inside a folder of the same name

File: src/Button/GreenButton.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import styled from 'styled-components'
+import Button from './Button'
+
+const GreenButton = styled(Button)`
```
Comment (jxnblk) [inline src/Button/GreenButton.js]: 👍

Comment (jxnblk) [reply src/Button/GreenButton.js]: Not to open a can of worms, but it might be worth sharing this pattern with your team and getting a sense of what they think of the naming convention here

File: src/Button/Button.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import styled from 'styled-components'
+import PropTypes from 'prop-types'
+import { fontSize, space } from 'styled-system'
+import { theme } from '../'
+
+const size = props => {
+  switch(props.size) {
+    case 'small':
```
Comment (jxnblk) [inline src/Button/Button.js]: This seems a lot more readable than the ternaries

File: src/Button/Button.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      return { 
+        height: '32px',
```
Comment (jxnblk) [inline src/Button/Button.js]: Fixed heights on buttons and inputs used to be sketchy with Firefox – make sure you do some monkey testing in a few browsers to make sure it looks okay

File: src/Button/Button.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  
+  &:hover {
+    background-color: ${props => props.theme.colors.blue1};
+  }
+
+  &:disabled {
+    opacity: 0.25;
+  }
+
+  width: ${props => props.fullWidth ? '100%' : 'initial' };
```
Comment (jxnblk) [inline src/Button/Button.js]: This is okay for now, but if you pull this out into its own function like `size`, you can avoid adding `width: initial` when `fullWidth` prop is falsy

File: src/Button/Button.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+  ${size}
```
Comment (jxnblk) [inline src/Button/Button.js]: Did you mean to add the `space` function here as well?

File: src/Button/OutlineButton.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import styled from 'styled-components'
+import PropTypes from 'prop-types'
+import Button from './Button'
+
+const OutlineButton = styled(Button)`
+  background-color: ${props => props.theme.colors.white};
```
Comment (jxnblk) [inline src/Button/OutlineButton.js]: I'd go with `background-color: transparent` here for support on non-white backgrounds

File: src/Button/OutlineButton.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  color: ${props => props.color || props.theme.colors.blue};
```
Comment (jxnblk) [inline src/Button/OutlineButton.js]: You could use a default prop here instead of the `||`

File: src/Button/OutlineButton.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  border-width: 2px;
+  border-color: ${props => props.color || props.theme.colors.blue};
```
Comment (jxnblk) [inline src/Button/OutlineButton.js]: In CSS, `border-color` is the same as the `color` for the element, so this line isn't needed. Same goes for if you were to refactor this with `boxShadow`

File: src/Button/OutlineButton.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+  &:hover {
+    background-color: ${props =>  props.theme.colors.white};
```
Comment (jxnblk) [inline src/Button/OutlineButton.js]: I thought the hover state was a dark blue color?

File: src/Button/OutlineButton.js (hunk 9 above)
Comment (jxnblk) [reply src/Button/OutlineButton.js]: I'd also either use the styled-system `color` function or a function that can pick up values from `theme.colors` by key since that's how the Box component works. E.g. the prop `color='blue'` gets the color `theme.colors.blue`

File: src/__tests__/Button.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    expect(json).toHaveStyleRule('font-size', '16px')
+  })
+
+  test('fullWidth prop sets width to 100%', () => {
+    const json = renderer.create(<Button fullWidth theme={theme} />).toJSON()
+    expect(json).toMatchSnapshot()
+    expect(json).toHaveStyleRule('width', '100%')
+  })
+
+  test('renders GreenButton', () => {
```
Comment (jxnblk) [inline src/__tests__/Button.js]: Maybe pull these into their own test files to mirror the `src` folder?

File: src/BackgroundImage.js (hunk 1 above)
Comment (hakimelek) [reply src/BackgroundImage.js]: I will rebase on master

File: src/Button/Button.js (hunk 2 above)
Comment (hakimelek) [reply src/Button/Button.js]: We can make it flat for now, but I think we will have to introduce some hierarchy later on in imo.

File: src/Button/Button.js (hunk 5 above)
Comment (hakimelek) [reply src/Button/Button.js]: Sure!

File: src/Button/Button.js (hunk 6 above)
Comment (hakimelek) [reply src/Button/Button.js]: So maybe just `${props => props.fullWidth ? 'width: 100%' : null}`

Comment (hakimelek) [reply src/Button/Button.js]: So maybe just `${props => props.fullWidth ? 'width: 100%' : null}`?

File: src/Button/Button.js (hunk 7 above)
Comment (hakimelek) [reply src/Button/Button.js]: I didn't try it, does it add `p`, `m`, `w` ?

File: src/Button/OutlineButton.js (hunk 9 above)
Comment (hakimelek) [reply src/Button/OutlineButton.js]: So maybe add `${color}` imported from `styled-system`?

File: src/Button/OutlineButton.js (hunk 11 above)
Comment (hakimelek) [reply src/Button/OutlineButton.js]: We need that yet!

File: src/Button/Button.js (hunk 6 above)
Comment (jxnblk) [reply src/Button/Button.js]: Yeah, I think that's a cleaner pattern to follow for styled-components

Comment (jxnblk) [reply src/Button/Button.js]: Might need the `;` in there too for it to compile though

File: src/Button/Button.js (hunk 7 above)
Comment (jxnblk) [reply src/Button/Button.js]: Yes, the `space` function adds these props (not `width` though): https://github.com/jxnblk/styled-system#space

File: src/Button/OutlineButton.js (hunk 11 above)
Comment (jxnblk) [reply src/Button/OutlineButton.js]: Maybe leave out the hover states and refactor after the colors are added then?

File: src/Button/GreenButton.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+  background-color: ${props => props.theme.colors.green};
+
+  &:hover {
+    background-color: ${props => props.theme.colors.green1};
+  }
+`
+export default GreenButton
```
Comment (BeniCheni) [inline src/Button/GreenButton.js]: Can this component become from <GreenButton /> to <ColorButton color='green' />? (thinking that maybe the button can be the colored button more dynamic instead of a fixed color; 
 same thought applies to the `RedButton`)

Comment (jxnblk) [reply src/Button/GreenButton.js]: We discussed this in person, but my suggestion was to start with explicitly named components, and abstract into props later if we need it. Part of the thinking is that having a component available is a little more obvious than props. 

Based on what I've seen in the product, we're mostly only using green and blue buttons, and by handling these as separate components we'll need less logic to handle hover, focus, and active states when those are added in.

File: src/BackgroundImage.js (hunk 1 above)
Comment (jxnblk) [reply src/BackgroundImage.js]: You should be able to pull the latest from master now and merge it into this branch. If you run into really hairy merge conflicts, I can help out

File: src/Button/Button.js (hunk 2 above)
Comment (jxnblk) [reply src/Button/Button.js]: Maybe, but we have no idea what that hierarchy should be yet – I'd say generally hold off on organizing things into folders until it becomes a development problem

**Comment by jxnblk** [comment]
Went ahead and pushed a few changes to get this merged in. We'll need to add hover states in once we have the full color scales in (#49), which can be handled in a separate PR
